zerver/lib/actions.py

```diff
@@ -5702,9 +5691,24 @@ def do_update_message(
             assert stream_being_edited is not None
             changed_msg_ids = [msg.id for msg in changed_messages]
 
-            if subs_gaining_usermessages:
+            if not new_stream.is_history_public_to_subscribers():
+                # For private streams, with history not public to subscribers,
+                # we find out users who didn't receive the message (by checking UserMessage rows)
+                # and create new UserMessage for these users so that they can
+                # access this message.
                 ums_to_create = []
+                ums_for_changed_msgs = UserMessage.objects.select_related().filter(
+                    message__id__in=changed_msg_ids
+                )
                 for msg_id in changed_msg_ids:
+                    user_ids_for_which_ums_exist = [
+                        um.user_profile.id for um in ums_for_changed_msgs if um.message_id == msg_id
```

Comment:
This ends up iterating through `ums_for_changed_msgs` many times, which is going to be really bad for performance.

Reply:
This fixes the issue:   

```
diff --git a/zerver/lib/actions.py b/zerver/lib/actions.py
index 2bce1306ef..da6bfb2eb5 100644
--- a/zerver/lib/actions.py
+++ b/zerver/lib/actions.py
@@ -5871,17 +5871,15 @@ def do_update_message(
                 # and create new UserMessage for these users so that they can
                 # access this message.


              ums_to_create = []
-                ums_for_changed_msgs = UserMessage.objects.select_related().filter(
-                    message__id__in=changed_message_ids
-                )
+                ums_by_message_id = defaultdict(set)
+                for um in UserMessage.objects.filter(message_id__in=changed_message_ids):
+                    ums_by_message_id[um.message_id].add(um.user_profile_id)
+
                 for msg_id in changed_message_ids:
-                    user_ids_for_which_ums_exist = [
-                        um.user_profile.id for um in ums_for_changed_msgs if um.message_id == msg_id
-                    ]
                     subs_gaining_usermessages = [
                         user_id
                         for user_id in new_stream_sub_ids
-                        if user_id not in user_ids_for_which_ums_exist
+                        if user_id not in ums_by_message_id[msg_id]
                     ]
                     for user_profile_id in subs_gaining_usermessages:
                         # The fact that the user didn't have a UserMessage originally means we can infer that the user
```